File: backend/app/api/v1/passages.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from app.database import get_db
from app.models.passage import Passage, PassageCategory, PassageDifficulty, PassageLanguage
from app.schemas.passage import PassageCreate, PassageResponse, PassageListResponse
from app.api.deps import get_current_user, get_admin_user
from app.models.user import User
from app.services.passage_service import passage_service
from typing import Optional, List
from uuid import UUID

router = APIRouter()


@router.get("/", response_model=List[PassageListResponse])
async def list_passages(
    category: Optional[PassageCategory] = None,
    difficulty: Optional[PassageDifficulty] = None,
    language: Optional[PassageLanguage] = None,
    limit: int = Query(default=50, le=100),
    offset: int = Query(default=0, ge=0),
    db: AsyncSession = Depends(get_db),
):
    passages = await passage_service.get_passages(db, category, difficulty, language, limit, offset)
    return passages


@router.get("/random", response_model=PassageResponse)
async def get_random_passage(
    category: PassageCategory = PassageCategory.SSC_CHSL,
    difficulty: PassageDifficulty = PassageDifficulty.MEDIUM,
    db: AsyncSession = Depends(get_db),
):
    passage = await passage_service.get_random_passage(db, category, difficulty)
    if not passage:
        raise HTTPException(status_code=404, detail="No passage found")
    return passage


@router.get("/{passage_id}", response_model=PassageResponse)
async def get_passage(passage_id: UUID, db: AsyncSession = Depends(get_db)):
    passage = await passage_service.get_passage(db, passage_id)
    if not passage:
        raise HTTPException(status_code=404, detail="Passage not found")
    return passage


@router.post("/", response_model=PassageResponse)
async def create_passage(
    data: PassageCreate,
    db: AsyncSession = Depends(get_db),
    admin: User = Depends(get_admin_user),
):
    passage = await passage_service.create_passage(db, data.model_dump())
    return passage

File: frontend/src/server/trpc/routers/passages.ts
import { router, publicProcedure } from '../trpc';
import { z } from 'zod';
import { passageService } from '../../services/passage-service';

export const passagesRouter = router({
  get: publicProcedure
    .input(z.object({ passageId: z.string() }))
    .query(async ({ input }) => {
      return passageService.getPassage(input.passageId);
    }),

  list: publicProcedure
    .input(z.object({
      category: z.string().optional(),
      difficulty: z.string().optional(),
      language: z.string().optional(),
      limit: z.number().min(1).max(100).default(50),
      offset: z.number().min(0).default(0),
    }))
    .query(async ({ input }) => {
      return passageService.getPassages(input);
    }),

  random: publicProcedure
    .input(z.object({
      category: z.string().optional(),
      difficulty: z.string().optional(),
      practiceSet: z.number().optional(),
    }))
    .query(async ({ input }) => {
      return passageService.getRandomPassage(input);
    }),

  create: publicProcedure
    .input(z.object({
      title: z.string().min(1),
      content: z.string().min(1),
      contentHindi: z.string().optional(),
      language: z.string().default('english'),
      category: z.string(),
      difficulty: z.string().default('medium'),
      exactKeyDepressions: z.number(),
      wordCount: z.number(),
      topic: z.string().optional(),
      source: z.string().optional(),
      sscExamYear: z.string().optional(),
      isExamLength: z.boolean().default(false),
    }))
    .mutation(async ({ input }) => {
      return passageService.createPassage(input);
    }),

  incrementUsage: publicProcedure
    .input(z.object({ passageId: z.string() }))
    .mutation(async ({ input }) => {
      await passageService.incrementUsage(input.passageId);
      return { success: true };
    }),
});
